source/xml_parser.c

```diff
@@ -0,0 +1,368 @@
+/*
+ * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/auth/private/xml_parser.h>
+
+#include <aws/common/array_list.h>
+
+static const size_t s_max_document_depth = 20;
+#define s_max_name_len ((size_t)256)
+#define s_node_close_overhead ((size_t)3)
+
+struct cb_stack_data {
+    aws_xml_parser_on_node_encountered_fn *cb;
+    void *user_data;
+};
+
+int aws_xml_parser_init(
+    struct aws_xml_parser *parser,
+    struct aws_allocator *allocator,
+    struct aws_byte_cursor *doc,
+    size_t max_depth) {
+    AWS_ZERO_STRUCT(*parser);
+    parser->allocator = allocator;
+    parser->doc = *doc;
+
+    parser->max_depth = s_max_document_depth;
+    parser->error = AWS_OP_SUCCESS;
+
+    if (max_depth) {
+        parser->max_depth = max_depth;
+    }
+
+    if (aws_array_list_init_dynamic(&parser->callback_stack, allocator, 4, sizeof(struct cb_stack_data))) {
+        return AWS_OP_ERR;
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+void aws_xml_parser_clean_up(struct aws_xml_parser *parser) {
+    if (parser->allocator) {
+        aws_array_list_clean_up(&parser->callback_stack);
+        AWS_ZERO_STRUCT(parser);
+    }
+}
+
+int s_node_next_sibling(struct aws_xml_parser *parser);
+
+static bool s_double_quote_fn(uint8_t value) {
+    return value == '"';
+}
+
+/* load the node declaration line, parsing node name and attributes.
+ *
+ * something of the form:
+ * <NodeName Attribute1=Value1 Attribute2=Value2 ...>
+ * */
+static int s_load_node_decl(
+    struct aws_xml_parser *parser,
+    struct aws_byte_cursor *decl_body,
+    struct aws_xml_node *node) {
+    struct aws_array_list splits;
+    AWS_ZERO_STRUCT(splits);
+
+    AWS_ZERO_ARRAY(parser->split_scratch);
+    aws_array_list_init_static(
+        &splits, parser->split_scratch, AWS_ARRAY_SIZE(parser->split_scratch), sizeof(struct aws_byte_cursor));
```

Comment:
If you want to avoid having this scratch space, you can use `aws_byte_cursor_next_split` in a loop to do a zero-memory-overhead split.

Reply:
I reallly like this idea, but since we have a limit of 11 splits already, it doesn't make a difference at the moment. I definitely want to revisit this when we do a full on public implementation without arbitrary limits.